api4/user.go

```diff
@@ -1129,6 +1156,69 @@ func updateUser(c *Context, w http.ResponseWriter, r *http.Request) {
 	w.Write([]byte(ruser.ToJson()))
 }
 
+// checkLoginProviderAttributes returns nil if the patch can be applied without overriding
+// attributes set by the user's login provider; otherwise, an error is returned whose text
+// is the name of the offending field.
+func checkLoginProviderAttributes(cfg *model.Config, ouser *model.User, patch *model.UserPatch) error {
+	type attrParam struct {
+		name          string
+		patchValue    *string
+		userValue     *string
+		providerValue *string
+	}
+	tryingToChange := func(patchValue *string, userValue *string) bool {
+		return patchValue != nil && *patchValue != *userValue
+	}
+	isConflict := func(patchValue *string, userValue *string, providerValue *string) bool {
+		return tryingToChange(patchValue, userValue) && *providerValue != ""
+	}
+	// If either the first name or last name is already set, then neither of them
+	// may be modified (LDAP and SAML users only)
+	isNameConflict := func(firstNameAttr *string, lastNameAttr *string) bool {
+		nameIsChanging := tryingToChange(patch.FirstName, &ouser.FirstName) ||
+			tryingToChange(patch.LastName, &ouser.LastName)
+		nameIsAlreadySet := *firstNameAttr != "" || *lastNameAttr != ""
+		return nameIsChanging && nameIsAlreadySet
+	}
+
+	if ouser.IsLDAPUser() {
+		LS := &cfg.LdapSettings
+		if isNameConflict(LS.FirstNameAttribute, LS.LastNameAttribute) {
+			return errors.New("full name")
+		}
+		for _, attr := range []attrParam{
+			{"nickname", patch.Nickname, &ouser.Nickname, LS.NicknameAttribute},
+			{"position", patch.Position, &ouser.Position, LS.PositionAttribute},
+			{"email", patch.Email, &ouser.Email, LS.EmailAttribute},
+		} {
+			if isConflict(attr.patchValue, attr.userValue, attr.providerValue) {
+				return errors.New(attr.name)
+			}
+		}
+	} else if ouser.IsSAMLUser() {
+		SS := &cfg.SamlSettings
+		if isNameConflict(SS.FirstNameAttribute, SS.LastNameAttribute) {
+			return errors.New("full name")
+		}
+		for _, attr := range []attrParam{
+			{"nickname", patch.Nickname, &ouser.Nickname, SS.NicknameAttribute},
+			{"position", patch.Position, &ouser.Position, SS.PositionAttribute},
+			{"email", patch.Email, &ouser.Email, SS.EmailAttribute},
+		} {
+			if isConflict(attr.patchValue, attr.userValue, attr.providerValue) {
+				return errors.New(attr.name)
+			}
+		}
+	}
```

Comment:
Don't we need to check isNameConflict for OAuth? I thought the prior code did.

Reply:
Yes, you're right. Completely forgot.